File: Java/456. 132 Pattern.java
class Solution {
    public boolean find132pattern(int[] nums) {
        Stack<Integer> stack = new Stack<>();
        int n = nums.length;
        int[] min = new int[n];
        min[0] = nums[0];
        
        for (int i=1; i<n; i++) {
            min[i] = Math.min(min[i-1], nums[i]);
        }
        
        for (int i=n-1; i>=0; i--) {
            if (nums[i] <= min[i]) {
                continue;
            }
            while (!stack.isEmpty() && stack.peek() <= min[i]) {
                stack.pop();
            }
            if (!stack.isEmpty() && stack.peek() < nums[i]) {
                return true;
            }
            stack.push(nums[i]);
        }
        
        return false;
    }
}

File: Python3/456. 132 Pattern.py
class Solution:
    def find132pattern(self, nums: List[int]) -> bool:
        stack = deque()
        min_arr = [nums[0]]
        for n in nums[1:]:
            min_arr.append(min(min_arr[-1], n))
                
        for i in range(len(nums)-1, -1, -1):
            if nums[i] <= min_arr[i]:
                continue
            
            while stack and stack[-1] <= min_arr[i]:
                stack.pop()
            if stack and stack[-1] < nums[i]:
                return True
            
            stack.append(nums[i])
            
        return False          